src/hotspot/share/opto/vectornode.hpp

```diff
@@ -1819,4 +1819,49 @@ class ExpandBitsVNode : public VectorNode {
   virtual int Opcode() const;
 };
 
+//------------------------------LoopVectorMaskNode------------------------------
+// Node for generating a loop vector mask from an integer range
+class LoopVectorMaskNode : public TypeNode {
+ private:
+  int _max_trips;
+ public:
+  LoopVectorMaskNode(Node* from, Node* to, const Type* ty, int max_trips) :
+      TypeNode(ty, 3), _max_trips(max_trips) {
+    init_class_id(Class_LoopVectorMask);
+    init_req(1, from);
+    init_req(2, to);
+  }
+
+  virtual int Opcode() const;
+  virtual uint hash() const { return TypeNode::hash() + _max_trips; }
+  virtual bool cmp(const Node& n) const {
+    return TypeNode::cmp(n) &&
+           _max_trips == ((LoopVectorMaskNode&)n)._max_trips;
```

Comment:
Is this cast really safe? Can you use `as_LoopVectorMaskNode()` instead, so we at least have an assert if it fails to be true? I fear this may get us into undefined behavior otherwise...

Reply:
Oh dear, I just saw the same pattern in:
```
bool TypeNode::cmp(const Node& n) const {
  return !Type::cmp(_type, ((TypeNode&)n)._type);
}
```
We should try to do more of that.

Reply:
Even if all callers currently ensure that `n` has the correct type, I'd say it is still not a great idea to cast without checking, at least in debug.

Reply:
I searched all C2 code and saw a lot of such patterns. Perhaps doing this in another RFE?

Reply:
Yes, please file an RFE. You can assign it to me. But for new code please already use `as_LoopVectorMaskNode()`.